cmd/thanos/rule.go

```diff
@@ -345,26 +365,47 @@ func runRule(
 		dns.ResolverType(conf.query.dnsSDResolver),
 	)
 	var (
-		queryClients []*httpconfig.Client
-		promClients  []*promclient.Client
+		queryClients    []*queryconfig.HTTPClient
+		promClients     []*promclient.Client
+		grpcEndpointSet *query.EndpointSet
+		grpcEndpoints   []string
 	)
+
 	queryClientMetrics := extpromhttp.NewClientMetrics(extprom.WrapRegistererWith(prometheus.Labels{"client": "query"}, reg))
+
 	for _, cfg := range queryCfg {
-		cfg.HTTPClientConfig.ClientMetrics = queryClientMetrics
-		c, err := httpconfig.NewHTTPClient(cfg.HTTPClientConfig, "query")
+		cfg.HTTPConfig.HTTPClientConfig.ClientMetrics = queryClientMetrics
+		c, err := queryconfig.NewHTTPClient(cfg.HTTPConfig.HTTPClientConfig, "query")
 		if err != nil {
 			return err
 		}
 		c.Transport = tracing.HTTPTripperware(logger, c.Transport)
-		queryClient, err := httpconfig.NewClient(logger, cfg.EndpointsConfig, c, queryProvider.Clone())
+		queryClient, err := queryconfig.NewClient(logger, cfg.HTTPConfig.EndpointsConfig, c, queryProvider.Clone())
 		if err != nil {
 			return err
 		}
 		queryClients = append(queryClients, queryClient)
 		promClients = append(promClients, promclient.NewClient(queryClient, logger, "thanos-rule"))
 		// Discover and resolve query addresses.
 		addDiscoveryGroups(g, queryClient, conf.query.dnsSDInterval)
+
+		if cfg.GRPCConfig != nil {
```

Comment:
My only concern is that `cfg.GRPCConfig` and `cfg.HTTPConfig` are both optional (https://github.com/thanos-io/thanos/pull/6390/files#diff-aa0502084cecb6911e7781e2b4fbd8cad71e1455cdfdad6c75a114ed6c7cc632R84), right? Maybe we can add the same check for `cfg.HTTPConfig` like `if cfg.HTTPConfig != nil { setup HTTP clients }`. This would make the code easier to read.

Reply:
I added a check for the HTTPConfig, which checks if there are not endpoints set. Is that what you had in mind?